File: lib/utils/announcement_category.dart
import 'package:flutter/material.dart';

import '../models/rt_unit_model.dart';
import '../widgets/warga_beranda_widgets.dart';

enum AnnouncementCategoryKind { kerjaBakti, info, darurat }

class AnnouncementCategoryMeta {
  final AnnouncementCategoryKind kind;
  final String label;
  final Color color;
  final Color bg;
  final Color dotColor;
  final IconData icon;

  const AnnouncementCategoryMeta({
    required this.kind,
    required this.label,
    required this.color,
    required this.bg,
    required this.dotColor,
    required this.icon,
  });
}

AnnouncementCategoryKind categoryKindFor(AnnouncementModel item) {
  final t = '${item.title} ${item.content}'.toLowerCase();
  if (t.contains('darurat') ||
      t.contains('peringatan') ||
      t.contains('bahaya') ||
      t.contains('pencurian') ||
      t.contains('emergency')) {
    return AnnouncementCategoryKind.darurat;
  }
  if (t.contains('kerja bakti') || t.contains('gotong royong') || t.contains('gotong')) {
    return AnnouncementCategoryKind.kerjaBakti;
  }
  if (item.isPinned && (t.contains('penting') || t.contains('waspada'))) {
    return AnnouncementCategoryKind.darurat;
  }
  return AnnouncementCategoryKind.info;
}

AnnouncementCategoryMeta categoryMetaFor(AnnouncementModel item) {
  switch (categoryKindFor(item)) {
    case AnnouncementCategoryKind.kerjaBakti:
      return const AnnouncementCategoryMeta(
        kind: AnnouncementCategoryKind.kerjaBakti,
        label: 'KERJA BAKTI',
        color: WargaBerandaColors.primaryGreen,
        bg: WargaBerandaColors.lightGreen,
        dotColor: WargaBerandaColors.primaryGreen,
        icon: Icons.event_rounded,
      );
    case AnnouncementCategoryKind.darurat:
      return const AnnouncementCategoryMeta(
        kind: AnnouncementCategoryKind.darurat,
        label: 'DARURAT',
        color: WargaBerandaColors.dangerRed,
        bg: WargaBerandaColors.accentRed,
        dotColor: WargaBerandaColors.dangerRed,
        icon: Icons.warning_amber_rounded,
      );
    case AnnouncementCategoryKind.info:
      return const AnnouncementCategoryMeta(
        kind: AnnouncementCategoryKind.info,
        label: 'INFO',
        color: Color(0xFF92400E),
        bg: WargaBerandaColors.accentYellow,
        dotColor: Color(0xFFEAB308),
        icon: Icons.campaign_rounded,
      );
  }
}

String announcementRelativeTime(DateTime createdAt) {
  final diff = DateTime.now().difference(createdAt);
  if (diff.inDays >= 7) {
    final weeks = (diff.inDays / 7).floor();
    return '$weeks minggu lalu';
  }
  if (diff.inDays >= 1) return '${diff.inDays} hari lalu';
  if (diff.inHours >= 1) return '${diff.inHours} jam lalu';
  if (diff.inMinutes >= 1) return '${diff.inMinutes} menit lalu';
  return 'Baru saja';
}

String announcementAuthorLabel(AnnouncementModel item) {
  final name = item.authorName?.trim();
  if (name == null || name.isEmpty) return 'Pengurus RT';
  if (name.toLowerCase().startsWith('pak ') || name.toLowerCase().startsWith('bu ')) {
    return name;
  }
  return 'Pak $name';
}

File: mobile/src/lib/announcementCategory.ts
// Port dari lib/utils/announcement_category.dart
import type { IconName } from '../components/Icon';
import { wargaColors } from '../config/theme';
import { Announcement } from '../types/models';

export type AnnouncementCategoryKind = 'kerjaBakti' | 'info' | 'darurat';

export interface AnnouncementCategoryMeta {
  kind: AnnouncementCategoryKind;
  label: string;
  color: string;
  bg: string;
  dotColor: string;
  icon: IconName;
}

export function categoryKindFor(item: Announcement): AnnouncementCategoryKind {
  const t = `${item.title} ${item.content}`.toLowerCase();
  if (
    t.includes('darurat') ||
    t.includes('peringatan') ||
    t.includes('bahaya') ||
    t.includes('pencurian') ||
    t.includes('emergency')
  ) {
    return 'darurat';
  }
  if (t.includes('kerja bakti') || t.includes('gotong royong') || t.includes('gotong')) {
    return 'kerjaBakti';
  }
  if (item.isPinned && (t.includes('penting') || t.includes('waspada'))) {
    return 'darurat';
  }
  return 'info';
}

export function categoryMetaFor(item: Announcement): AnnouncementCategoryMeta {
  switch (categoryKindFor(item)) {
    case 'kerjaBakti':
      return {
        kind: 'kerjaBakti',
        label: 'KERJA BAKTI',
        color: wargaColors.primaryGreen,
        bg: wargaColors.lightGreen,
        dotColor: wargaColors.primaryGreen,
        icon: 'calendar',
      };
    case 'darurat':
      return {
        kind: 'darurat',
        label: 'DARURAT',
        color: wargaColors.dangerRed,
        bg: wargaColors.accentRed,
        dotColor: wargaColors.dangerRed,
        icon: 'warning',
      };
    case 'info':
    default:
      return {
        kind: 'info',
        label: 'INFO',
        color: '#92400E',
        bg: wargaColors.accentYellow,
        dotColor: '#EAB308',
        icon: 'megaphone',
      };
  }
}

export function announcementRelativeTime(createdAt: Date): string {
  const diffMs = Date.now() - createdAt.getTime();
  const mins = Math.floor(diffMs / 60000);
  const hours = Math.floor(mins / 60);
  const days = Math.floor(hours / 24);
  if (days >= 7) return `${Math.floor(days / 7)} minggu lalu`;
  if (days >= 1) return `${days} hari lalu`;
  if (hours >= 1) return `${hours} jam lalu`;
  if (mins >= 1) return `${mins} menit lalu`;
  return 'Baru saja';
}

export function announcementAuthorLabel(item: Announcement): string {
  const name = item.authorName?.trim();
  if (!name) return 'Pengurus RT';
  const lower = name.toLowerCase();
  if (lower.startsWith('pak ') || lower.startsWith('bu ')) return name;
  return `Pak ${name}`;
}
